src/modules/dashboard/components/ActionsPage/ActionsPage.tsx

```diff
@@ -299,6 +303,55 @@ const ActionsPage = () => {
     tokenInfo: { decimals, symbol },
   } = tokenData;
 
+  const getFormmatedRoleList = (roleGroupA, roleGroupB) => {
+    let roleList = '';
+
+    roleGroupA.forEach((role, i) => {
+      const roleNameMessage = { id: `role.${role.id}` };
+      const formattedRole = formatMessage(roleNameMessage);
+
+      roleList += ` ${formattedRole}`;
+
+      if (
+        i < roleGroupA.length - 1 ||
+        (i === roleGroupA.length - 1 && !isEmpty(roleGroupB))
+      ) {
+        roleList += ',';
+      }
+    });
+
+    return roleList;
+  };
+
+  const getRolesTitle = () => {
+    let roleTitle = '';
+    const assignedRoles = roles.filter((role) => role.setTo);
+    const unassignedRoles = roles.filter((role) => !role.setTo);
+
+    if (!isEmpty(assignedRoles)) {
+      roleTitle += `assigned${getFormmatedRoleList(
+        assignedRoles,
+        unassignedRoles,
+      )}`;
+    }
+
+    if (isEmpty(assignedRoles) && !isEmpty(unassignedRoles)) {
+      roleTitle += `removed${getFormmatedRoleList(
+        unassignedRoles,
+        assignedRoles,
+      )}`;
+    } else if (!isEmpty(unassignedRoles)) {
+      roleTitle += ` and removed the${getFormmatedRoleList(
+        unassignedRoles,
+        assignedRoles,
+      )}`;
+    }
+
+    roleTitle += roles.length > 1 ? ' permissions' : ' permission';
+
+    return roleTitle;
+  };
+
```

Comment:
i would extract those methods to some utils/helper file, or utils/roles

Reply:
I second this. They have no palce inside a React component